lib/auth/grpcserver_test.go

```diff
@@ -0,0 +1,301 @@
+	as, err := NewTestAuthServer(TestAuthServerConfig{
+		Dir:   t.TempDir(),
+		Clock: clock,
+	})
+	require.NoError(t, err)
+	srv, err := as.NewTestTLSServer()
+	require.NoError(t, err)
+	// Enable U2F support.
+	authPref, err := services.NewAuthPreference(types.AuthPreferenceSpecV2{
+		Type: teleport.Local,
+		U2F: &types.U2F{
+			AppID:  "teleport",
+			Facets: []string{"teleport"},
+		},
+	})
+	require.NoError(t, err)
+	err = as.AuthServer.SetAuthPreference(authPref)
+	require.NoError(t, err)
+
+	// Create a fake user.
+	user, _, err := CreateUserAndRole(as.AuthServer, "mfa-user", []string{"role"})
+	require.NoError(t, err)
+	cl, err := srv.NewClient(TestUser(user.GetName()))
+	require.NoError(t, err)
+
+	// No MFA devices should exist for a new user.
+	resp, err := cl.GetMFADevices(ctx, &proto.GetMFADevicesRequest{})
+	require.NoError(t, err)
+	require.Empty(t, resp.Devices)
+
+	totpSecrets := make(map[string]string)
+	u2fDevices := make(map[string]*mocku2f.Key)
+
+	// Add several MFA devices.
+	addTests := []struct {
+		desc string
+		opts mfaAddTestOpts
+	}{
+		{
+			desc: "add initial TOTP device",
+			opts: mfaAddTestOpts{
+				initReq: &proto.AddMFADeviceRequestInit{
+					DeviceName: "totp-dev",
+					Type:       proto.AddMFADeviceRequestInit_TOTP,
+				},
+				authHandler: func(t *testing.T, req *proto.MFAAuthenticateChallenge) *proto.MFAAuthenticateResponse {
+					// The challenge should be empty for the first device.
+					require.Empty(t, cmp.Diff(req, &proto.MFAAuthenticateChallenge{}))
+					return &proto.MFAAuthenticateResponse{}
+				},
+				registerHandler: func(t *testing.T, req *proto.MFARegisterChallenge) *proto.MFARegisterResponse {
+					totpRegisterChallenge := req.GetTOTP()
+					require.NotEmpty(t, totpRegisterChallenge)
+					require.Equal(t, totpRegisterChallenge.Algorithm, otp.AlgorithmSHA1.String())
+					code, err := totp.GenerateCodeCustom(totpRegisterChallenge.Secret, clock.Now(), totp.ValidateOpts{
+						Period:    uint(totpRegisterChallenge.Period),
+						Digits:    otp.Digits(totpRegisterChallenge.Digits),
+						Algorithm: otp.AlgorithmSHA1,
+					})
+					require.NoError(t, err)
+
+					totpSecrets["totp-dev"] = totpRegisterChallenge.Secret
+					return &proto.MFARegisterResponse{
+						Response: &proto.MFARegisterResponse_TOTP{TOTP: &proto.TOTPRegisterResponse{
+							Code: code,
+						}},
+					}
+				},
+				wantDev: func(t *testing.T) *types.MFADevice {
+					wantDev, err := services.NewTOTPDevice("totp-dev", totpSecrets["totp-dev"], clock.Now())
+					require.NoError(t, err)
+					return wantDev
+				},
+			},
+		},
+		{
+			desc: "add a U2F device",
+			opts: mfaAddTestOpts{
+				initReq: &proto.AddMFADeviceRequestInit{
+					DeviceName: "u2f-dev",
+					Type:       proto.AddMFADeviceRequestInit_U2F,
+				},
+				authHandler: func(t *testing.T, req *proto.MFAAuthenticateChallenge) *proto.MFAAuthenticateResponse {
+					// Respond to challenge using the existing TOTP device.
```

Comment:
iiuc, responding to a challenge with existing devices with no response (e.g. like in the first test-case w/o any existing devices) should also be considered a failure? I'm not sure how and whether this is enforced though

Reply:
didn't understand the comment, are you suggesting to add a negative test case which fails the auth challenge?

Reply:
Yes, I meant that the flow is hinting that when user has already registered a device, it requires its token as a response and this is not enforced in the test. I'm not sure whether this is a concern since this depends on how the client is implemented.

Reply:
Added negative test cases for failing auth and registration challenges for both add and delete